micrometer-core/src/main/java/io/micrometer/core/instrument/binder/kafka/KafkaMetrics.java

```diff
@@ -0,0 +1,279 @@
+                    }
+                }
+            });
+        }
+    }
+
+    @NotNull private Meter bindMeter(MeterRegistry registry, Metric metric) {
+        String metricName = metricName(metric);
+        Meter meter;
+        if (metricName.endsWith("total") || metricName.endsWith("count")) {
+            meter = registerCounter(registry, metric, metricName, extraTags);
+        } else if (metricName.endsWith("min")
+                || metricName.endsWith("max")
+                || metricName.endsWith("avg")) {
+            meter = registerGauge(registry, metric, metricName, extraTags);
+        } else if (metricName.endsWith("rate")) {
+            meter = registerTimeGauge(registry, metric, metricName, extraTags);
+        } else {
+            meter = registerGauge(registry, metric, metricName, extraTags);
+        }
+        return meter;
+    }
+
+    private TimeGauge registerTimeGauge(MeterRegistry registry, Metric metric, String metricName, Iterable<Tag> extraTags) {
+        return TimeGauge.builder(metricName, metric, TimeUnit.SECONDS, toMetricValue(registry))
+            .tags(metricTags(metric))
+            .tags(extraTags)
+            .description(metric.metricName().description())
+            .register(registry);
+    }
+
+    private Gauge registerGauge(MeterRegistry registry, Metric metric, String metricName, Iterable<Tag> extraTags) {
+        return Gauge.builder(metricName, metric, toMetricValue(registry))
+            .tags(metricTags(metric))
+            .tags(extraTags)
+            .description(metric.metricName().description())
+            .register(registry);
+    }
+
+    private FunctionCounter registerCounter(MeterRegistry registry, Metric metric, String metricName, Iterable<Tag> extraTags) {
+        return FunctionCounter.builder(metricName, metric, toMetricValue(registry))
+            .tags(metricTags(metric))
+            .tags(extraTags)
+            .description(metric.metricName().description())
+            .register(registry);
+    }
+
+    private ToDoubleFunction<Metric> toMetricValue(MeterRegistry registry) {
+        return metric -> {
+            //Double-check if new metrics are registered; if not (common scenario)
+            //it only adds metrics count validation
+            checkAndBindMetrics(registry);
+            if (metric.metricValue() instanceof Double) {
+                return (double) metric.metricValue();
+            } else {
+                return 0.0;
+            }
+        };
+    }
+
+    private List<Tag> metricTags(Metric metric) {
+        return metric.metricName().tags()
+            .entrySet()
+            .stream()
+            .map(entry -> Tag.of(entry.getKey(), entry.getValue()))
+            .collect(Collectors.toList());
+    }
+
+    private String metricName(Metric metric) {
+        String value =
+            METRIC_NAME_PREFIX + metric.metricName().group() + "." + metric.metricName().name();
```

Comment:
I didn't check which part it is coming from but I noticed `metrics` is included in every metric name. This feels superfluous and just makes the name longer. What do you think about getting rid of the "metrics" part of the name. Unless of course there are some kafka metrics about metrics. Or if Kafka users would expect the name with the `metrics` part in it, then maybe it is better to keep it. What do you think?

Reply:
I agree it looks much better without `-metrics` suffix in between. I don't think there is any real issue other than people used to or having dashboards that include metrics with this suffix included, as comes from JMX. If we're ok with prioritizing better naming than familiarity/similarity with JMX naming, I can update this.

Reply:
I was also thinking about compatibility with the previous `KafkaConsumerMetrics` as Micrometer users may already have dashboards built around those and it would be best if the consumer metrics could continue to be used as is with the new implementation.